pkg/component/extensions/operatingsystemconfig/controlplanebootstrap.go

```diff
@@ -0,0 +1,160 @@
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"k8s.io/utils/ptr"
+	"sigs.k8s.io/controller-runtime/pkg/client"
+
+	gardencorev1beta1 "github.com/gardener/gardener/pkg/apis/core/v1beta1"
+	v1beta1constants "github.com/gardener/gardener/pkg/apis/core/v1beta1/constants"
+	extensionsv1alpha1 "github.com/gardener/gardener/pkg/apis/extensions/v1alpha1"
+	"github.com/gardener/gardener/pkg/component/extensions/operatingsystemconfig/nodeinit"
+	"github.com/gardener/gardener/pkg/controllerutils"
+	"github.com/gardener/gardener/pkg/extensions"
+)
+
+type controlPlaneBootstrap struct {
+	// This type only implements a subset of Interface. Embed this to spare the panic stubs.
+	Interface
+
+	log    logr.Logger
+	client client.Client
+	values *ControlPlaneBootstrapValues
+
+	waitInterval        time.Duration
+	waitSevereThreshold time.Duration
+	waitTimeout         time.Duration
+
+	osc *extensionsv1alpha1.OperatingSystemConfig
+}
+
+// ControlPlaneBootstrapValues contains the values used to create an OperatingSystemConfig resource for gardenadm bootstrap.
+type ControlPlaneBootstrapValues struct {
+	// Namespace is the namespace for the OperatingSystemConfig resource.
+	Namespace string
+	// Worker is the control plane worker pool.
+	Worker *gardencorev1beta1.Worker
+	// GardenadmImage is the gardenadm image reference that should be pulled.
+	GardenadmImage string
+}
+
+// NewControlPlaneBootstrap creates a new instance of Interface for deploying OperatingSystemConfigs during gardenadm bootstrap.
+func NewControlPlaneBootstrap(
+	log logr.Logger,
+	client client.Client,
+	values *ControlPlaneBootstrapValues,
+	waitInterval, waitSevereThreshold, waitTimeout time.Duration,
+) Interface {
+	return &controlPlaneBootstrap{
+		log:    log,
+		client: client,
+		values: values,
+
+		waitInterval:        waitInterval,
+		waitSevereThreshold: waitSevereThreshold,
+		waitTimeout:         waitTimeout,
+
+		osc: &extensionsv1alpha1.OperatingSystemConfig{ObjectMeta: metav1.ObjectMeta{
+			Name:      "gardenadm-" + values.Worker.Name,
+			Namespace: values.Namespace,
+		}},
+	}
+}
+
+func (c *controlPlaneBootstrap) Deploy(ctx context.Context) error {
+	units, files, err := nodeinit.GardenadmConfig(c.values.GardenadmImage)
+	if err != nil {
+		return err
+	}
+
+	_, err = controllerutils.GetAndCreateOrMergePatch(ctx, c.client, c.osc, func() error {
+		metav1.SetMetaDataAnnotation(&c.osc.ObjectMeta, v1beta1constants.GardenerOperation, v1beta1constants.GardenerOperationReconcile)
+		metav1.SetMetaDataAnnotation(&c.osc.ObjectMeta, v1beta1constants.GardenerTimestamp, TimeNow().UTC().Format(time.RFC3339Nano))
+
+		c.osc.Spec = extensionsv1alpha1.OperatingSystemConfigSpec{
+			Purpose: extensionsv1alpha1.OperatingSystemConfigPurposeProvision,
+			Units:   units,
+			Files:   files,
+		}
+
+		if c.values.Worker.Machine.Image != nil {
+			c.osc.Spec.Type = c.values.Worker.Machine.Image.Name
+			c.osc.Spec.ProviderConfig = c.values.Worker.Machine.Image.ProviderConfig
+		}
```

Comment:
We should fail (somewhere?) if the worker machine image is not set in the high-touch scenario. Otherwise, we don't know which OS extension to use, and the OSC will never get reconciled.
There was a similar reference to this field when deploying the `Worker`. But we don't run any validation in gardenadm yet, so I just assumed it's always set.
I think it should be fine to omit the nil check for now and rely on the to-be-introduced Shoot validation.

Reply:
Done in 8ba7d4e395139c0b2841b27aec039ef5f5a1adbd